src/mjolnir/speed_assigner.h

```diff
@@ -0,0 +1,353 @@
+    if (!config_file) {
+      LOG_INFO("Disabled default speeds assignment from config");
+      return;
+    }
+    try {
+      auto doc = rapidjson::read_json(*config_file);
+      if (doc.HasParseError())
+        throw std::runtime_error("malformed json");
+      if (!doc.IsArray())
+        throw std::runtime_error("must be a json array");
+
+      // loop over each country/state pair
+      for (const auto& cs : doc.GetArray()) {
+        std::string code;
+        if (cs.HasMember("iso3166-1")) {
+          code += cs["iso3166-1"].GetString();
+          if (code.empty())
+            throw std::runtime_error("Cannot have empty country code");
+          if (cs.HasMember("iso3166-2")) {
+            code.push_back('.');
+            if (code.size() == (code += cs["iso3166-2"].GetString()).size())
+              throw std::runtime_error("Cannot have emtpy state code");
+          }
+        }
+        if (tables.count(code))
+          throw std::runtime_error("Duplicate country/state entry");
+        tables.emplace(std::move(code), std::array<SpeedTable, 3>{
+                                            SpeedTable(cs["rural"]),
+                                            SpeedTable(cs["suburban"]),
+                                            SpeedTable(cs["urban"]),
+                                        });
+      }
+    } // something went wrong with parsing or opening the file
+    catch (const std::exception& e) {
+      LOG_WARN(std::string("Disabled default speeds assignment from config: ") + e.what());
+      tables.clear();
+    } // something else was thrown
+    catch (...) {
+      LOG_WARN("Disabled default speeds assignment from config: unknown error");
+      tables.clear();
+    }
+    LOG_INFO("Enabled default speeds assignment from config: " + *config_file);
+  }
+
+  /**
+   * Update directed edge speed based on density and other edge parameters like
+   * surface type. TODO - add admin specific logic
+   * @param directededge         Directed edge to update.
+   * @param density              Relative road density.
+   * @param rc_speed             Array of default speeds vs. road class
+   * @param infer_turn_channels  Flag indicating if turn channels were inferred
+   * @param country_code         2 letter country code
+   * @param state_code           2 letter state code
+   * @return  true if config based speeds were used to update the edge.
+   *          false indicates the conventional heuristic based approach was used
+   */
+  bool UpdateSpeed(DirectedEdge& directededge,
+                   const uint32_t density,
+                   bool infer_turn_channels,
+                   const std::string& country_code,
+                   const std::string& state_code) const {
+
+    // If we have config loaded we'll use that
+    if (!tables.empty() && FromConfig(directededge, density, country_code, state_code)) {
+      directededge.set_speed_type(SpeedType::kClassified);
```

Comment:
@rzyc i added this since you last reviewed realizing that it would be set improperly if the speed was a tagged speed